### PR TITLE
Add XCode support for C++ executables and libraries

### PR DESCRIPTION
**Pull request by lacasseio**
### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=dl%2Fxcode-ide-cpp)
- [x] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [x] Recognize contributor in release notes

### CODE REVIEW

**Comment by adammurdoch** [comment]
I think we need a little more test coverage. We should make sure the generation works well when there is a transitive dependency in the graph, eg app -> lib1 -> lib2. In particular, we should make sure that the header search path for each project is correct.

**Comment by big-guy** [comment]
Could we have some simple infrastructure and smoke tests to run xcodebuild on the generated files only on macOS?

**Comment by lacasseio** [comment]
@adammurdoch I added more tests and unroll the existing ones. Could you give some feedback on those.

@big-guy I like the idea, I will look into implementing them tomorrow.

File: subprojects/ide-native/src/integTest/groovy/org/gradle/ide/xcode/XcodeMultiProjectIntegrationTest.groovy [hunk 1]
```diff
@@ -43,27 +43,202 @@ rootProject.name = "${WORKSPACE_NAME}"
...
-        and:
-        workspaceXml.FileRef.size() == 3
-        workspaceXml.FileRef*.@location*.replaceAll('absolute:', '').containsAll([file('app/app.xcodeproj'), file('greeter/greeter.xcodeproj')]*.absolutePath)
+        xcodeWorkspace("${rootProjectName}.xcworkspace")
+            .contentFile.assertHasProjects([file("${rootProjectName}.xcodeproj"), file('app/app.xcodeproj'), file('greeter/greeter.xcodeproj')]*.absolutePath)
+
+        buildSettings(xcodeProject("app/app.xcodeproj").projectFile).HEADER_SEARCH_PATHS == toSpaceSeparatedList(file("app/src/main/headers"), file("greeter/src/main/public"))
+    }
+
+    def "Gradle project with added xcode plugin are not included in the workspace"() {
```
Comment (adammurdoch) [inline subprojects/ide-native/src/integTest/groovy/org/gradle/ide/xcode/XcodeMultiProjectIntegrationTest.groovy]: Is this actually the behaviour we want? if so, should verify that the include path is still configured correctly.

I would add this test coverage as a later change, so we can get the changes we already have merged into master.

File: subprojects/ide-native/src/integTest/groovy/org/gradle/ide/xcode/XcodeMultiProjectIntegrationTest.groovy [hunk 2]
```diff
@@ -16,24 +16,24 @@
...
+import org.gradle.nativeplatform.fixtures.app.ExeWithLibraryUsingSwiftLibraryHelloWorldApp
 import org.gradle.nativeplatform.fixtures.app.SwiftHelloWorldApp
+import org.gradle.util.CollectionUtils
 
-class XcodeMultiProjectIntegrationTest extends AbstractIntegrationSpec {
-    private static final String WORKSPACE_NAME = "workspace"
-    def app = new SwiftHelloWorldApp()
+import static org.gradle.ide.xcode.internal.XcodeUtils.toSpaceSeparatedList
+
+class XcodeMultiProjectIntegrationTest extends AbstractXcodeIntegrationSpec {
```
Comment (adammurdoch) [inline subprojects/ide-native/src/integTest/groovy/org/gradle/ide/xcode/XcodeMultiProjectIntegrationTest.groovy]: the single project tests are separated by language, whereas they are all grouped into a single class for multi-project tests. I would separate the multi-projects by language. Do this as a later change so we can get the current changes into mater.

**Comment by adammurdoch** [comment]
@lacasseio if you do add the xcodebuild verification, please do this as another PR after this one is merged. The changes in this PR are ready to go and we shouldn't delay them by adding more scope.

File: subprojects/ide-native/src/integTest/groovy/org/gradle/ide/xcode/XcodeMultiProjectIntegrationTest.groovy (hunk 1 above)
Comment (lacasseio) [inline subprojects/ide-native/src/integTest/groovy/org/gradle/ide/xcode/XcodeMultiProjectIntegrationTest.groovy]: Oups, my bad. It's actually the opposite and that is what the test is actually doing. I will fix the test name to reflect what the test is actually testing.

**Pull request merged**